pkg/queue/concurrency_state.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+// ConcurrencyStateHandler tracks the in flight requests for the pod. When the requests
+// drop to zero, it runs the `pause` function, and when requests scale up from zero, it
+// runs the `resume` function. If either of `pause` or `resume` are not passed, it runs
+// the respective local function(s). The local functions are the expected behavior; the
+// function parameters are enabled primarily for testing purposes.
+func ConcurrencyStateHandler(endpoint string, h http.Handler, logger *zap.SugaredLogger, pause, resume func()) http.HandlerFunc {
+	logger.Info("Concurrency state tracking enabled")
+
+	if pause == nil {
+		pause = func() {
```

Comment:
I'd probably pull this out to a separate top-level function or interface implementation rather than defining the default inline here. That should also be a lot easier to test (which we should do :)).